attest/tpm_linux.go

```diff
@@ -118,281 +89,16 @@ func openTPM(tpm probedTPM) (*TPM, error) {
-		if err != nil {
-			return nil, fmt.Errorf("CertifyCreation failed: %v", err)
-		}
-		// Pack the raw structure into a TPMU_SIGNATURE.
-		signature, err := tpmutil.Pack(tpm2.AlgRSASSA, tpm2.AlgSHA256, tpmutil.U16Bytes(sig))
-		if err != nil {
-			return nil, fmt.Errorf("failed to pack TPMT_SIGNATURE: %v", err)
-		}
-		return &AK{ak: newKey20(keyHandle, blob, pub, creationData, attestation, signature)}, nil
-	default:
-		return nil, fmt.Errorf("unsupported TPM version: %x", t.version)
-	}
-}
-
-func (t *linuxTPM) loadAK(opaqueBlob []byte) (*AK, error) {
-	sKey, err := deserializeKey(opaqueBlob, t.version)
-	if err != nil {
-		return nil, fmt.Errorf("deserializeKey() failed: %v", err)
-	}
-	if sKey.Encoding != keyEncodingEncrypted {
-		return nil, fmt.Errorf("unsupported key encoding: %x", sKey.Encoding)
-	}
-
-	switch sKey.TPMVersion {
-	case TPMVersion12:
-		return &AK{ak: newKey12(sKey.Blob, sKey.Public)}, nil
-	case TPMVersion20:
-		srk, _, err := t.getPrimaryKeyHandle(commonSrkEquivalentHandle)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get SRK handle: %v", err)
-		}
-		var hnd tpmutil.Handle
-		if hnd, _, err = tpm2.Load(t.rwc, srk, "", sKey.Public, sKey.Blob); err != nil {
-			return nil, fmt.Errorf("Load() failed: %v", err)
-		}
-		return &AK{ak: newKey20(hnd, sKey.Blob, sKey.Public, sKey.CreateData, sKey.CreateAttestation, sKey.CreateSignature)}, nil
-	default:
-		return nil, fmt.Errorf("cannot load AK with TPM version: %v", sKey.TPMVersion)
-	}
-}
-
-// allPCRs12 returns a map of all the PCR values on the TPM
-func allPCRs12(ctx *tspi.Context) (map[uint32][]byte, error) {
-	tpm := ctx.GetTPM()
-	PCRlist, err := tpm.GetPCRValues()
-	if err != nil {
-		return nil, fmt.Errorf("failed to read PCRs: %v", err)
-	}
-
-	PCRs := make(map[uint32][]byte)
-	for i := 0; i < len(PCRlist); i++ {
-		PCRs[(uint32)(i)] = PCRlist[i]
-	}
-	return PCRs, nil
-}
-
-func (t *linuxTPM) pcrs(alg HashAlg) ([]PCR, error) {
-	var PCRs map[uint32][]byte
-	var err error
-
-	switch t.version {
-	case TPMVersion12:
-		if alg != HashSHA1 {
-			return nil, fmt.Errorf("non-SHA1 algorithm %v is not supported on TPM 1.2", alg)
-		}
-		PCRs, err = allPCRs12(t.ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to read PCRs: %v", err)
-		}
-
-	case TPMVersion20:
-		PCRs, err = readAllPCRs20(t.rwc, alg.goTPMAlg())
-		if err != nil {
-			return nil, fmt.Errorf("failed to read PCRs: %v", err)
-		}
-
-	default:
-		return nil, fmt.Errorf("unsupported TPM version: %x", t.version)
-	}
-
-	out := make([]PCR, len(PCRs))
-	for index, digest := range PCRs {
-		out[int(index)] = PCR{
-			Index:     int(index),
-			Digest:    digest,
-			DigestAlg: alg.cryptoHash(),
-		}
-	}
-
-	return out, nil
-}
 
-func (t *linuxTPM) measurementLog() ([]byte, error) {
-	return ioutil.ReadFile("/sys/kernel/security/tpm0/binary_bios_measurements")
+	panic("unreachable")
```

Comment:
This isn't unreachable right? It should return an error about being passed an unrecognized TPM version

Reply:
It shouldnt be possible to set `version` to anything other than those two values without using unsafe (see code above, which populates `probedTPM.version`)

Reply:
Wait wut

Reply:
Sec

Reply:
Done. What the code said and what I thought it said were very different things ^^'